sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingQueryListenerBus.scala

```diff
@@ -70,6 +70,10 @@ class StreamingQueryListenerBus(sparkListenerBus: LiveListenerBus)
         sparkListenerBus.post(s)
         // post to local listeners to trigger callbacks
         postToAll(s)
+      case t: QueryTerminatedEvent =>
+        // run all the listeners synchronized before removing the id from the list
+        postToAll(t)
```

Comment:
It will post to all listeners directly in the current thread. Hence, the listener may see QueryProgressEvent after QueryTerminatedEvent.

You can override `postToAll`. It's fine to remove `final` from the `postToAll` method.